Lib/functools.py

```diff
@@ -826,3 +826,32 @@ def wrapper(*args, **kw):
     wrapper._clear_cache = dispatch_cache.clear
     update_wrapper(wrapper, func)
     return wrapper
+
+
+################################################################################
+### cached_property() - computed once per instance, cached as attribute
+################################################################################
+
+class cached_property:
+    def __init__(self, func):
+        self.func = func
+        self.__doc__ = func.__doc__
+        self.lock = RLock()
+
+    def __get__(self, instance, cls=None):
+        if instance is None:
+            return self
+        attrname = self.func.__name__
+        try:
+            cache = instance.__dict__
+        except AttributeError:  # objects with __slots__ have no __dict__
+            msg = (
+                f"No '__dict__' attribute on {type(instance).__name__!r} "
+                f"instance to cache {attrname!r} property."
+            )
+            raise TypeError(msg) from None
+        with self.lock:
```

Comment:
`__getitem__` / `setdefault` are atomic operations.
The code could be modified in the following way:
1. Get attribute value *without* locking.
2. If cached attr is not found -- acquire the lock, get the value again under the lock and calculate it if not present.

Reply:
Sure, will update. Thanks for the review!